src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs

```diff
@@ -70,8 +69,9 @@ protected override Task DisposeCoreAsync(bool initialized)
 
         internal async Task OnProjectChangedAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> e = null)
         {
-            bool isDebuggable = await _launchProviders.Value.IsDebuggableAsync()
-                                                            .ConfigureAwait(false);
+            bool isDebuggable = await _launchProviders.Value.IsDebuggableAsync();
+
+            await JoinableFactory.SwitchToMainThreadAsync();
```

Comment:
Indeed. So is my reasoning correct, and this just happens to be the exception that proves the rule?  The threading brownbag, though a few years old, mentioned that there was no way to tell what is UI bound or not. Is that still true?

Reply:
cc @drewnoakes 

The unwritten rule is:

- All IVsXXX are touched and QI'd (any casts) on the UI thread, *unless* you've literally spoken to the team or API is documented. 

More and more APIs in VS are becoming free-threaded & thread-safe - especially those used during solution load/reload/branch switch or other perf sensitive areas.

Reply:
With regards to the cast, here's a bug I just wrote the day:

``` C#
        private async Task<IVsSolution> GetSolutionAsync()
        {
			return (IVsSolution)await _serviceProvider.GetServiceAsync(typeof(SVsSolution)).
        }
```

That cast is a COM QueryInterface under the covers and because IVsSolution's implementation is an STA object, this results an RPC call if I'm not on the same thread it was created on. RPC is a blocking call, which can lead to [thread-pool exhaustion](https://github.com/Microsoft/vs-threading/blob/master/doc/threadpool_starvation.md) so we try and avoid them. Naively pushing code from UI thread -> background thread can easily result in performance getting worse, not better.

In this project, we have IVsUIService<T> and IVsService<T> interfaces that can simplify/enforce usage of VS services. IVsUIService<T> is used for services that are UI-thread based, and IVsService<T> for services that are can be used on background threads.

Reply:
> IVsService for services that are can be used on background threads

So the fact that `_startupProjectsListService` is an `IVsService<IVsStartupProjectsListService>` is what tells you that the switch to main thread is unnecessary? That makes it very clear, thank you.

Reply:
Assuming someone didn't use it incorrect, yes. I've thought about writing a analyzer that catches usage of IVsService for interfaces that aren't on a known set of "safe" instances.

Reply:
Is there a link for the threading brownbag?

Reply:
Under the Team's OneNote - "Getting started with CPS" section.